### PR TITLE
Allow setting http headers and custom http status codes with FastAPI GraphQLRouter

### PR DESCRIPTION
**Pull request by davidnemec**
## Description

Title says it all. I need to set response headers/cookies with FastAPI GraphQLRouter.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Probably #1416

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to David Němec for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Review by patrick91 — COMMENTED** [review]
Looks good, do you think we can also add support for status code and cookies?

**Comment by davidnemec** [comment]
Cookies are also supported by this (they are just headers too). It's even included in tests.

I can add support for status codes if that's something desired (I needed it few months ago but at the end I solved it without using custom http status). Can I do that in separate PR so we don't block merging this?

**Comment by patrick91** [comment]
> Cookies are also supported by this (they are just headers too). It's even included in tests.
Wasn't sure how fastapi handled cookies, cool 😊
 
> I can add support for status codes if that's something desired (I needed it few months ago but at the end I solved it without using custom http status). Can I do that in separate PR so we don't block merging this?
Let's do that in another PR then 😊

File: strawberry/fastapi/router.py [hunk 1]
```diff
@@ -208,6 +218,14 @@ def get_graphiql_response(self) -> HTMLResponse:
         html = get_graphiql_html()
         return HTMLResponse(html)
 
+    @staticmethod
+    def __merge_responses(response: Response, actual_response: Response) -> Response:
```
Comment (patrick91) [inline strawberry/fastapi/router.py:222]:
```suggestion
    def _merge_responses(response: Response, actual_response: Response) -> Response:
```

File: strawberry/fastapi/router.py [hunk 2]
```diff
@@ -157,7 +163,11 @@ async def handle_http_query(
 
             response_data = await self.process_result(request, result)
 
-            return JSONResponse(response_data, status_code=status.HTTP_200_OK)
+            actual_response = JSONResponse(
+                response_data,
+                status_code=status.HTTP_200_OK,
+            )
+            return self.__merge_responses(response, actual_response)
```
Comment (patrick91) [inline strawberry/fastapi/router.py:170]:
```suggestion
            return self._merge_responses(response, actual_response)
```

File: strawberry/fastapi/router.py [hunk 3]
```diff
@@ -134,18 +138,20 @@ async def handle_http_query(
...
 
             try:
                 request_data = parse_request_data(data)
             except MissingQueryError:
-                return PlainTextResponse(
+                actual_response = PlainTextResponse(
                     "No GraphQL query found in the request",
                     status_code=status.HTTP_400_BAD_REQUEST,
                 )
+                return self.__merge_responses(response, actual_response)
```
Comment (patrick91) [inline strawberry/fastapi/router.py:154]:
```suggestion
                return self._merge_responses(response, actual_response)
```

File: strawberry/fastapi/router.py [hunk 4]
```diff
@@ -134,18 +138,20 @@ async def handle_http_query(
                     operations, files_map, multipart_data
                 )
             else:
-                return PlainTextResponse(
+                actual_response = PlainTextResponse(
                     "Unsupported Media Type",
                     status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
                 )
+                return self.__merge_responses(response, actual_response)
```
Comment (patrick91) [inline strawberry/fastapi/router.py:145]:
```suggestion
                return self._merge_responses(response, actual_response)
```

File: strawberry/fastapi/router.py [hunk 5]
```diff
@@ -122,10 +125,11 @@ async def handle_http_query(
                 try:
                     data = await request.json()
                 except json.JSONDecodeError:
-                    return PlainTextResponse(
+                    actual_response = PlainTextResponse(
                         "Unable to parse request body as JSON",
                         status_code=status.HTTP_400_BAD_REQUEST,
                     )
+                    return self.__merge_responses(response, actual_response)
```
Comment (patrick91) [inline strawberry/fastapi/router.py:132]:
```suggestion
                    return self._merge_responses(response, actual_response)
```

File: tests/fastapi/test_response_status.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    app = FastAPI()
+    schema = strawberry.Schema(query=Query)
+    graphql_app = GraphQLRouter(schema)
+    app.include_router(graphql_app, prefix="/graphql")
+
+    test_client = TestClient(app)
+    response = test_client.post("/graphql", json={"query": "{ abc }"})
+
+    assert response.status_code == 418
```
Comment (patrick91) [inline tests/fastapi/test_response_status.py:27]: 🫖

**Review by patrick91 — APPROVED** [review]
Looks great! I've only requested a change (I'm going to commit it) regarding to how we mark a function as private, I don't think we use `__` anywhere in the codebase. 😊

**Comment by davidnemec** [comment]
> Looks great! I've only requested a change (I'm going to commit it) regarding to how we mark a function as private, I don't think we use `__` anywhere in the codebase. blush

Thanks. I was not sure about that one :slightly_smiling_face:

**Comment by davidnemec** [comment]
Thanks for fixing mypy issue I was looking into in too.

**Comment by patrick91** [comment]
> Thanks for fixing mypy issue I was looking into in too.

ah sorry! should have told you 😊 going to merge this as soon as all the check pass! thanks for the PR <3

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**